[APM] Add Uptime link to Transaction action menu

dgieselaar opened the pull request:
Closes #35144

Adds a link to the transaction menu to view a transaction (sample) in Uptime.

## Summary

~~Label is still WIP, have to consult w/ Uptime about what the right label is (and if the URL is correct).~~

Moves TransactionActionMenu to hooks, and makes sure there's only one branch where the element for a link gets created (instead of two like before). Small change, but will help if there are any changes to the layout.

From

![image](https://user-images.githubusercontent.com/352732/57468659-33fdd200-7285-11e9-9470-aafc52ef1a2c.png)


To 

![image](https://user-images.githubusercontent.com/352732/57468619-1d577b00-7285-11e9-9b9c-92e55e2c065b.png)


### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- ~~[ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)~~
- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- ~~[ ] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials~~
- ~~[ ] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios~~
- ~~[ ] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)~~

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)
- [ ] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)

dgieselaar committed:
[APM] Add Uptime link to Transaction action menu

Adds a link to the transaction menu to view a transaction (sample) in Uptime.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13381/)

formgeist commented:
@makwarth Any suggestion what the label should be? "View status"? We don't normally include the app name in the label, but perhaps it's needed here?

makwarth commented:
@formgeist perhaps "View monitor status". The icon should indicate Uptime.

dgieselaar committed:
Change label, make link optional, pass domain as search query to Uptime

- Change label to "View monitor status"
- Only show Uptime link when transaction.url is available
- Pass domain as search query to Uptime

formgeist commented:
@makwarth SGTM

formgeist commented on x-pack/plugins/apm/public/components/shared/TransactionActionMenu/TransactionActionMenu.tsx:
```diff
@@ -43,168 +49,184 @@ function ActionMenuButton({ onClick }: { onClick: () => void }) {
+      child: (
+        <EuiLink href={uptimeLink}>
+          {i18n.translate('xpack.apm.transactionActionMenu.viewInUptime', {
+            defaultMessage: 'View in Uptime'
```
Change default message to: "View monitor status"

formgeist replied:
@dgieselaar You're too fast 😄

dgieselaar replied:
@formgeist haha 🚀🚀🚀

sqren commented on x-pack/plugins/apm/public/components/shared/TransactionActionMenu/TransactionActionMenu.tsx:
```diff
@@ -43,168 +49,189 @@ function ActionMenuButton({ onClick }: { onClick: () => void }) {
+  items: Array<T | null | undefined>
+): T[] {
+  return items.filter(item => !!item) as T[];
+}
```
Why not keep the inline filter on the `condition`?

dgieselaar replied:
TypeScript doesn't understand what filter does (AFAICT) without a little help. This is a little nudge for TypeScript to understand that there will not be any null or undefined items in the array. Let's see what it looks like after I use `condition` everywhere though.

dgieselaar replied:
No longer necessary 👍

sqren commented on x-pack/plugins/apm/public/components/shared/TransactionActionMenu/TransactionActionMenu.tsx:
```diff
@@ -43,168 +49,189 @@ function ActionMenuButton({ onClick }: { onClick: () => void }) {
+  const podId = idx(transaction, _ => _.kubernetes.pod.uid);
+  const containerId = idx(transaction, _ => _.container.id);
+  const traceId = idx(transaction, _ => _.trace.id);
+  const time = new Date(transaction['@timestamp']).getTime();
```
Better to use `transaction.timestamp.us`

sqren replied:
(note, this will be in microseconds instead of the current millis)

dgieselaar replied:
Sure! FWIW, just lines moved around.

sqren commented on x-pack/plugins/apm/public/components/shared/TransactionActionMenu/TransactionActionMenu.tsx:
```diff
@@ -43,168 +49,189 @@ function ActionMenuButton({ onClick }: { onClick: () => void }) {
+export const TransactionActionMenu: FunctionComponent<Props> = (
+  props: Props
+) => {
+  const { transaction } = props;
```
nit: This is a matter of personal preference but what about:
```ts
export TransactionActionMenu({ transaction }: Props) {
...
}
```

There are some issues in Dev tools and jest snapshots where the component name is omitted when exporting an an (anonymous) arrow function instead of a named function.
Also, the destructuring of props seems nice in this case. 

Ofc. this is subjective and matter of style - not important stuff :)

dgieselaar replied:
Do you mean `export function TransactionMenu`? Is there a way I can reproduce it? Just want to check if that is still the case.. I think Babel at least should handle this appropriately because it is a named export (but not sure).

sqren replied:
Yes, I missed the `function` part 😅

sqren replied:
It happens when using enzyme's `mount` afair. Will check again.

dgieselaar replied:
I've left this for now - a `function` export is a little harder to type as a FunctionalComponent, which means that you lose some type safety (AFAIK). I'll look into the mount stuff later, could not find anything about it quickly.

sqren commented on x-pack/plugins/apm/public/components/shared/TransactionActionMenu/TransactionActionMenu.tsx:
```diff
@@ -43,168 +49,189 @@ function ActionMenuButton({ onClick }: { onClick: () => void }) {
+            )
+          )}`
+        })
+      : null;
```
I think it's cleaner to add it to the items unconditionally, and then set the `condition` like:
```js
condition: transaction && transaction.url
```

dgieselaar replied:
Will fix.

dgieselaar replied:
Fixed.

sorenlouv reviewed (approved):
Apart from a few nits 👍

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13496/)

dgieselaar committed:
Use condition everywhere, fix TS error, use timestamp.us

- Use `condition` on all newly created action menu items, then filter out any falsy conditions
- Use timestamp.us instead of costlier new Date(transaction['@timestamp'])
- Fix TypeScript error in es fieldnames test

sqren commented on x-pack/plugins/apm/public/components/shared/TransactionActionMenu/TransactionActionMenu.tsx:
```diff
@@ -47,164 +53,180 @@ interface Props {
+            )
+          )}`
+        })
+      : null;
```
What do you think about getting rid of the ternary and `pick`:

```js
const uptimeLink = url.format({
  pathname: chrome.addBasePath('/app/uptime'),
  hash: `/?${fromQuery({
    dateRangeStart: urlParams.rangeFrom,
    dateRangeEnd: urlParams.rangeTo,
    search: `url.domain:${idx(transaction, _ => _.url.domain)}`
  })}`
});
```

And below:
```js
condition: transaction && transaction.url
```

This way it's obvious what the condition for showing the link is when glancing over the list of items.

dgieselaar replied:
`pick` is in there to prevent empty URL parameters ending up in the URL (any key in the object will be added to the URL, whether it's `null`, `undefined` or an empty string).

Agree w/ the other comment, looks cleaner (1 condition instead of 2).

sqren replied:
Yeah, I don't mind that but @ogupte mentioned that too so I'll give in :)

dgieselaar committed:
Always construct uptimeLink, move condition to item itself for more consistency

dgieselaar commented:
@justinkambic Link is now, as an example, `http://localhost:5601/qwi/app/uptime#/?dateRangeStart=now-24h&dateRangeEnd=now&search=url.domain%3Aopbeans-node`. Does that match your expectation?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13518/)

sorenlouv reviewed (approved):
LGMT - thanks for doing the cleanup. Much nicer than before!

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13528/)

justinkambic commented:
> @justinkambic Link is now, as an example, http://localhost:5601/qwi/app/uptime#/?dateRangeStart=now-24h&dateRangeEnd=now&search=url.domain%3Aopbeans-node. Does that match your expectation?

Yes! 👍

sqren commented:
@dgieselaar One thing: You should add `Closes #35144` at the top of the PR description. Then the issue is automatically closed when this PR is merged, and it's easier to correlate the two :)

justinkambic reviewed (approved):
LGTM

sqren commented on x-pack/plugins/apm/public/components/shared/TransactionActionMenu/TransactionActionMenu.tsx:
```diff
@@ -47,164 +53,177 @@ interface Props {
+          })}
+        </EuiLink>
+      ),
+      condition: transaction && transaction.url
```
@dgieselaar Just noticed this. Shouldn't it be `idx(transaction, t => t.url.domain)`?
